README.md
[![Build Status](https://travis-ci.org/Yomguithereal/dolman.svg)](https://travis-ci.org/Yomguithereal/dolman)

# Dolman

**dolman** is a very light wrapper around an express app meaning to provide some comfortable utilities for one developing a REST API.

As such, dolman overload express' response object to provide handy methods such as `.ok` or `.notFound` and provides a easy way to deploy routers using straightforward configuration.

# Summary

* [Installation](#installation)
* [Usage](#usage)
  * [Wrapping an express app](#wrapping-an-express-app)
  * [Responses](#responses)
  * [Router](#router)
  * [Specs](#specs)
* [What on earth is a dolman?](#explanation)
* [License](#license)

## Installation

You can easily install **dolman** using `npm`:

```bash
npm install --save dolman
```

## Usage

### Wrapping an express app

```js
var express = require('express'),
    wrap = require('dolman');

var app = express();

// Wrapping the express app & providing some options:
var dolman = wrap(app);

// Example with a custom typology:
var dolman = wrap(app, {typology: myCustomTypology});

// Example with a custom logger:
var dolman = wrap(app, {
  logger: {
    error: function(msg, data) {},
    warn: ...
  }
});
```

### Responses

A wrapped express app will have an enhanced response object:

```js
var app = express(),
    dolman = wrap(app);

app.get('/hello', function(req, res) {
  return res.ok({hello: 'world'});
});

// The following methods are available:
res.ok([result]);
res.created([result]);
res.badRequest([reason]);
res.unauthorized();
res.forbidden();
res.notFound([reason]);
res.serverError([error]);
```

Note that the responses will be sent as JSON and will follow this standard:

```js
// Typical success
{
  status: 'ok',
  code: 200,
  result: {
    hello: 'world'
  }
}

// Typical error
{
  status: 'error',
  code: 400,
  error: {
    message: 'Bad Request',
    reason: {
      source: 'body',
      expecting: {
        id: 'number'
      },
      sent: {
        id: 'whatever'
      }
    }
  }
}
```

### Router

**dolman** exposes an easy way to create express routers by using lists of configuration objects representing a single route.

```js
var app = express(),
    dolman = wrap(app);

// Creating a router with a single route
var router = dolman.router([
  {
    url: '/hello',
    action: function(req, res) {
      return res.ok({hello: req.query.name});
    }
  }
]);

// Creating the same router with beforehand authentication
var router = dolman.router(authMiddleware, routes);

app.use(router);
// or to use a namespace
app.use('/prefix', router);
```

A route can be described likewise:

```js
{
  // [required] - The matching express pattern.
  // Remember that this could even be an array or a regex.
  url: '/hello/:name',

  // [required] - The action to perform.
  action: function(req, res) {
    return res.ok({hello: req.params.name});
  },

  // You can alternatively pass an array of functions
  action: [fn1, fn2],

  // [optional] - The route's name.
  name: 'hello',

  // [optional] - The route's description.
  description: 'Say hello to someone.',

  // [optional] - The accepted methods.
  // If not provided, defaults to 'ALL'
  method: 'POST',

  // The following also works:
  methods: ['POST', 'PUT'],

  // [optional] - Validation for params, query and body.
  validate: {
    params: {
      name: 'string'
    },
    query: {
      age: '?number'
    },
    body: {
      title: 'string',
      content: {
        id: '?number',
        text: 'string'
      }
    }
  },

  // [optional] - Cache specifications
  cache: 'hello',

  // Or, if the cached response is relative to the request's parameters:
  cache: {
    key: 'hello',
    hasher: function(req) {
      return req.params.name;
    }
  },

  // [optional] You can leverage HTTP Cache-Control headers as well.
  httpCache: 'no-cache, no-store, must-revalidate',

  // Or, shorthand and fluent way to set 'private, max-age':
  httpCache: {
    hours: 2 // seconds, minutes, hours, days, weeks
  },

  // [optional] - A typology definition used to check and clamp the data output.
  mask: {
    name: 'string',
    age: '?number'
  }
}
```

Note that the validation specifications are handled by the [typology](https://github.com/jacomyal/typology) library.

## Specs

**dolman** gives you the opportunity to retrieve very easily the specifications of your API if you need them, for instance, to hydrate a client elsewhere.

The output will follow the [SPORE](https://github.com/SPORE/specifications) specifications.

```js
var express = require('express'),
    wrap = require('dolman');

var app = express(),
    dolman = wrap(express);

var router = dolman.router([
  {
    url: '/hello',
    name: 'hello',
    description: 'Say hello.',
    action: function(req, res) {
      return res.ok({hello: 'world'});
    }
  }
]);

app.use('/nested', router);

dolman.specs();
>>> {
  formats: ['json'],
  methods: {
    hello: {
      path: '/nested/hello',
      name: 'hello',
      description: 'Say hello.'
    }
  }
}
```

Note that **dolman** will only output your routes having a name.

<h2 id="explanation">What on earth is a dolman?</h2>

A [dolman](https://en.wikipedia.org/wiki/Dolman) is a loose garment with narrow sleeves, an opening in the front and lavish braids. While originating from Turkey, this garment was mostly worn as a jacket by hussars.

## License

[MIT](LICENSE.txt)


CHANGELOG.md
# Changelog

## 0.4.0

* Dolman now wraps the express app rather than the express dependency (breaking).
* Adding the `logger` option.
* Adding `mask`.
* Changing the arity of `dolman.specs`.

## 0.3.1

* Fixing `typology` dependency's scheme.

## 0.3.0

* Adding possibility to pass multiple functions as action.
* Adding possibility to pass middlewares before routes.
* Adding method field to `dolman.specs` output.

## 0.2.0

* Adding `name` and `description`.
* Adding `dolman.specs`.

## 0.1.0

* Adding `httpCache` support.


test/unit.js
/**
 * Dolman Unit Tests
 * ==================
 *
 * The library's unit tests.
 */
var Typology = require('typology'),
    express = require('express'),
    bodyParser = require('body-parser'),
    request = require('supertest'),
    assert = require('assert'),
    sinon = require('sinon'),
    async = require('async'),
    wrap = require('../index.js');

var customTypes = new Typology({
  custom: function(v) {
    return v === 'custom';
  }
});

/**
 * Responses.
 */
describe('Responses', function() {
  var RESPONSES = [
    {method: 'ok', code: 200},
    {method: 'created', code: 201},
    {method: 'badRequest', code: 400},
    {method: 'unauthorized', code: 401},
    {method: 'forbidden', code: 403},
    {method: 'notFound', code: 404},
    {method: 'serverError', code: 500}
  ];

  it('express\'s responses should have been enhanced.', function() {
    var app = express();

    wrap(app);

    RESPONSES.forEach(function(response) {
      assert(typeof app.response[response.method] === 'function');
    });
  });

  it('an app should return the correct codes.', function(done) {
    var app = express();

    wrap(app);

    RESPONSES.forEach(function(response) {

      app.get('/' + response.method, function(req, res) {
        return res[response.method]();
      });
    });

    async.eachSeries(RESPONSES, function(response, next) {

      request(app)
        .get('/' + response.method)
        .expect(response.code, next);
    }, done);
  });
});

/**
 * Router.
 */
describe('Router', function() {
  var dolman,
      app;

  beforeEach(function() {
    app = express();
    dolman = wrap(app, {typology: customTypes});
  });

  describe('Basics', function() {
    it('a route without url should throw.', function() {
      assert.throws(function() {
        dolman.router([
          {
            action: function(req, res) {
              return res.send('hello');
            }
          }
        ]);
      }, /no url/);
    });

    it('a route without action should throw.', function() {
      assert.throws(function() {
        dolman.router([
          {
            url: '/hello'
          }
        ]);
      }, /no action/);
    });

    it('basic routes should work.', function(done) {
      var router = dolman.router([
        {
          url: '/test',
          action: function(req, res) {
            return res.ok();
          }
        }
      ]);

      app.use(router);

      request(app)
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(200, {
          code: 200,
          status: 'ok'
        }, done);
    });

    it('method(s) should work.', function(done) {
      var action = function(req, res) {
        return res.ok();
      };

      var router = dolman.router([
        {
          url: '/post',
          method: 'POST',
          action: action
        },
        {
          url: '/lowercase',
          method: 'post',
          action: action
        },
        {
          url: '/put',
          methods: ['POST', 'PUT'],
          action: action
        }
      ]);

      app.use(router);

      async.series({
        notFound: function(next) {
          request(app)
            .get('/post')
            .expect(404, next);
        },
        post: function(next) {
          request(app)
            .post('/post')
            .expect(200, next);
        },
        lowercase: function(next) {
          request(app)
            .post('/lowercase')
            .expect(200, next);
        },
        put: function(next) {
          request(app)
            .put('/put')
            .expect(200, next);
        }
      }, done);
    });

    it('passing multiple actions should work.', function(done) {
      var router = dolman.router([
        {
          url: '/hello',
          action: [
            function(req, res, next) {
              req.message = {hello: 'world'};
              return next();
            },
            function(req, res) {
              return res.json(req.message);
            }
          ]
        }
      ]);

      app.use(router);

      request(app)
        .get('/hello')
        .expect(200, {hello: 'world'}, done);
    });

    it('should be possible to pass series of middlewares before the routes.', function(done) {
      var authName = function(req, res, next) {
        if (req.query.name === 'John')
          return next();
        return res.forbidden();
      };

      var authSurname = function(req, res, next) {
        if (req.query.surname === 'Black')
          return next();
        return res.notFound();
      };

      var router = dolman.router(authName, authSurname, [
        {
          url: '/hello',
          action: function(req, res) {
            return res.json({hello: 'world'});
          }
        }
      ]);

      app.use(router);

      async.series({
        rejectedName: function(next) {
          request(app)
            .get('/hello')
            .expect(403, next);
        },
        rejectedSurname: function(next) {
          request(app)
            .get('/hello?name=John')
            .expect(404, next);
        },
        accepted: function(next) {
          request(app)
            .get('/hello?name=John&surname=Black')
            .expect(200, {hello: 'world'}, next);
        }
      }, done);
    });
  });

  describe('Validation', function() {
    it('validation middleware should work.', function(done) {

      var controller = [
        {
          url: '/url/:id',
          validate: {
            params: {
              id: 'custom'
            },
            query: {
              title: 'string'
            }
          },
          method: 'GET',
          action: function(req, res) {
            return res.ok();
          }
        },
        {
          url: '/url/:id',
          validate: {
            params: {
              id: 'custom'
            },
            body: {
              text: 'string',
              number: '?number'
            }
          },
          method: 'POST',
          action: function(req, res) {
            return res.ok({hello: req.body.text});
          }
        }
      ];

      app.use(bodyParser.json());

      app.use(dolman.router(controller));

      async.series({
        invalidParams: function(next) {
          request(app)
            .get('/url/shawarma')
            .expect(400, {
              code: 400,
              status: 'error',
              error: {
                message: 'Bad Request',
                reason: {
                  source: 'params',
                  path: '/url/:id',
                  expecting: {
                    id: 'custom'
                  },
                  sent: {
                    id: 'shawarma'
                  }
                }
              }
            }, next);
        },
        invalidQuery: function(next) {
          request(app)
            .get('/url/custom')
            .expect(400, {
              code: 400,
              status: 'error',
              error: {
                message: 'Bad Request',
                reason: {
                  source: 'query',
                  expecting: {
                    title: 'string'
                  },
                  sent: {}
                }
              }
            }, next);
        },
        invalidBody: function(next) {
          request(app)
            .post('/url/custom')
            .send({number: 42})
            .expect(400, {
              code: 400,
              status: 'error',
              error: {
                message: 'Bad Request',
                reason: {
                  source: 'body',
                  expecting: {
                    text: 'string',
                    number: '?number'
                  },
                  sent: {
                    number: 42
                  }
                }
              }
            }, next);
        },
        valid: function(next) {
          request(app)
            .post('/url/custom')
            .send({text: 'world'})
            .expect(200, {
              code: 200,
              status: 'ok',
              result: {
                hello: 'world'
              }
            }, next);
        }
      }, done);
    });
  });

  describe('Cache', function() {
    it('RAM cache middleware should work.', function(done) {
      var one = 0,
          two = 0;

      var router = dolman.router([
        {
          url: '/one',
          cache: 'one',
          action: function(req, res) {
            one++;
            return res.ok({hello: 'world'});
          }
        },
        {
          url: '/two',
          cache: {
            key: 'two',
            hasher: function(req) {
              return req.query.title;
            }
          },
          action: function(req, res) {
            two++;
            return res.ok({title: req.query.title});
          }
        }
      ]);

      app.use(router);

      async.series({
        one: function(next) {
          request(app)
            .get('/one')
            .expect(200, {
              code: 200,
              status: 'ok',
              result: {
                hello: 'world'
              }
            }, next);
        },
        oneCached: function(next) {
          request(app)
            .get('/one')
            .expect(200, {
              code: 200,
              status: 'ok',
              result: {
                hello: 'world'
              }
            }, next);
        },
        two: function(next) {
          request(app)
            .get('/two?title=world')
            .expect(200, {
              code: 200,
              status: 'ok',
              result: {
                title: 'world'
              }
            }, next);
        },
        twoCached: function(next) {
          request(app)
            .get('/two?title=world')
            .expect(200, {
              code: 200,
              status: 'ok',
              result: {
                title: 'world'
              }
            }, next);
        },
        count: function(next) {
          assert(one === 1);
          assert(two === 1);
          return next();
        }
      }, done);
    });

    it('HTTP cache middleware should work.', function(done) {
      var val1 = 'public, max-age=3600',
          val2 = 'private, max-age=59', // 59 seconds
          val3 = 'private, max-age=60', // 1 minute
          val4 = 'private, max-age=7200', // 2 hours
          val5 = 'private, max-age=259200', // 3 days
          val6 = 'private, max-age=2419200'; // 4 weeks

      var action = function(req, res) {
        return res.ok();
      };

      var router = dolman.router([
        {
          url: '/string',
          httpCache: val1,
          action: action
        },
        {
          url: '/seconds',
          httpCache: {seconds: 59},
          action: action
        },
        {
          url: '/minutes',
          httpCache: {minutes: 1},
          action: action
        },
        {
          url: '/hours',
          httpCache: {hours: 2},
          action: action
        },
        {
          url: '/days',
          httpCache: {days: 3},
          action: action
        },
        {
          url: '/weeks',
          httpCache: {weeks: 4},
          action: action
        }
      ]);

      app.use(router);

      async.series({
        stringParam: function(next) {
          request(app)
            .get('/string')
            .expect('Cache-Control', val1, next);
        },
        seconds: function(next) {
          request(app)
            .get('/seconds')
            .expect('Cache-Control', val2, next);
        },
        minutes: function(next) {
          request(app)
            .get('/minutes')
            .expect('Cache-Control', val3, next);
        },
        hours: function(next) {
          request(app)
            .get('/hours')
            .expect('Cache-Control', val4, next);
        },
        days: function(next) {
          request(app)
            .get('/days')
            .expect('Cache-Control', val5, next);
        },
        weeks: function(next) {
          request(app)
            .get('/weeks')
            .expect('Cache-Control', val6, next);
        }
      }, done);
    });

    it('HTTP cache middle should throw an error on unexpected params.', function(done) {
      var router;

      function setup() {
        router = dolman.router([
          {
            url: '/foo',
            httpCache: {no: 'way'},
            action: function(req, res) {
              return res.ok();
            }
          }
        ]);
        app.use(router);
      }

      assert.throws(setup, Error);
      request(app).get('/foo').end(done);
    });
  });

  describe('Masks', function() {

    it('masks should clamp sent data.', function(done) {
      var router = dolman.router([
        {
          url: '/masked',
          mask: {
            one: 'number',
            two: 'number'
          },
          action: function(req, res) {
            return res.ok({
              one: 1,
              two: 2,
              three: 3
            });
          }
        }
      ]);

      app.use(router);

      request(app)
        .get('/masked')
        .expect(200, {
          status: 'ok',
          code: 200,
          result: {
            one: 1,
            two: 2
          }
        }, done);
    });

    it('masks should also clamp collections.', function(done) {
      var router = dolman.router([
        {
          url: '/masked',
          mask: [{name: 'string', age: 'number'}],
          action: function(req, res) {
            return res.ok([
              {
                name: 'John',
                surname: 'White',
                age: 45
              },
              {
                name: 'Jack',
                surname: 'Black',
                age: 56
              }
            ]);
          }
        }
      ]);

      app.use(router);

      request(app)
        .get('/masked')
        .expect(200, {
          status: 'ok',
          code: 200,
          result: [
            {
              name: 'John',
              age: 45
            },
            {
              name: 'Jack',
              age: 56
            }
          ]
        }, done);
    });

    it('masks should warn the user when sent data is not valid.', function(done) {
      var spy = sinon.spy(),
          logger = {warn: spy, error: spy};

      app = express();
      dolman = wrap(app, {logger: logger});

      var router = dolman.router([
        {
          url: '/primitive',
          mask: 'number',
          action: function(req, res) {
            return res.ok('hello');
          }
        }
      ]);

      app.use(router);

      request(app)
        .get('/primitive')
        .expect(200, function() {
          assert(spy.called);
          done();
        });
    });
  });
});

/**
 * Specifications.
 */
describe('Specifications', function() {

  it('should be possible to retrieve our app\'s specifications.', function() {
    var app = express(),
        dolman = wrap(app);

    var router1 = dolman.router([
      {
        url: '/hello',
        name: 'hello',
        description: 'Say hello.',
        action: function(req, res) {
          return res.ok({hello: 'world'});
        }
      },
      {
        url: '/greet/:name',
        name: 'greet',
        method: 'POST',
        action: function(req, res) {
          return res.ok({hello: req.params.name});
        }
      }
    ]);

    var router2 = dolman.router([
      {
        url: '/goodbye',
        name: 'goodbye',
        methods: ['PUT', 'POST'],
        description: 'Good Bye.',
        action: function(req, res) {
          return res.ok({goodbye: 'world'});
        }
      }
    ]);

    app.use(router1);
    app.use('/nested', router2);

    var specs = dolman.specs();

    assert.deepEqual(specs, {
      formats: ['json'],
      methods: {
        hello: {
          path: '/hello',
          name: 'hello',
          description: 'Say hello.',
          method: 'GET'
        },
        greet: {
          path: '/greet/:name',
          name: 'greet',
          method: 'POST'
        },
        goodbye: {
          path: '/nested/goodbye',
          name: 'goodbye',
          description: 'Good Bye.',
          method: 'PUT'
        }
      }
    });
  });
});


test/helpers.js
/**
 * Dolman Test Helpers
 * ====================
 *
 * Miscellaneous helper functions.
 */
var express = require('express'),
    dolman = require('../index.js')(express);

function makeApp() {
  var app = express();
}

module.exports = {
  app: makeApp
};


index.js
/**
 * Dolman API Endpoint
 * ====================
 *
 * Exposing the library's utilities.
 */
var responses = require('./src/responses.js'),
    createLogger = require('./src/createLogger.js'),
    middlewares = require('./src/middlewares.js'),
    helpers = require('./src/helpers.js'),
    express = require('express'),
    Typology = require('typology'),
    join = require('path').join,
    util = require('util');

module.exports = function(app, opts) {
  opts = opts || {};

  // Creating the logger
  var logger = opts.logger;

  if (!logger)
    logger = createLogger('logger' in opts && !opts.logger);

  // Building internal typology
  var types;

  if (opts.typology instanceof Typology)
    types = opts.typology;
  else
    types = new Typology(opts.typology || {});

  // Applying responses to the express app
  responses(app, logger, types);

  // Internal route register
  var routesMap = new Map();

  // Internal RAM cache
  var cache = {};

  /**
   * Router function.
   */
  function makeRouter() {

    // Solving arguments
    var args = [].slice.call(arguments);

    var beforeMiddlewares = args.slice(0, -1),
        routes = args[args.length - 1];

    var router = express.Router();

    routes.forEach(function(route) {
      if (!route.url)
        throw Error('dolman.router: one route has no url: ' + util.inspect(route));

      if (!route.action)
        throw Error('dolman.router: the route for url ' + route.url + ' has no action.');

      var actions = [].concat(route.action);

      // Storing the route
      routesMap.set(actions[0], route);

      // Applying before middlewares
      var routeMiddlewares = beforeMiddlewares.slice();

      // Validation
      if (route.validate)
        routeMiddlewares.push(middlewares.validate(types, route.validate));

      // Mask
      if (route.mask)
        routeMiddlewares.push(middlewares.mask(route.mask));

      // RAM cache
      if (route.cache)
        routeMiddlewares.push(middlewares.cache(cache, route.cache));

      // HTTP cache
      if (route.httpCache)
        routeMiddlewares.push(middlewares.httpCache(route.httpCache));

      // Determining the method
      var methods = [].concat(route.method || route.methods || 'ALL');

      methods.forEach(function(method) {
        router[method.toLowerCase()].apply(
          router,
          [route.url]
            .concat(routeMiddlewares)
            .concat(actions)
        );
      });
    });

    return router;
  }

  /**
   * Specifications functions.
   */
  function specs() {
    var routes = {};

    // Reducing the app's recursive stack
    function reduceStack(path, items, item) {
      var nextPath;

      if (item.handle && item.handle.stack) {
        nextPath = join(path, (item.path || helpers.unescapeRegex(item.regexp) || ''));
        return items.concat(item.handle.stack.reduce(reduceStack.bind(null, nextPath), []));
      }

      if (item.route) {
        nextPath = join(path, (item.route.path || ''));
        return items.concat(item.route.stack.reduce(reduceStack.bind(null, nextPath), []));
      }

      return items.concat({
        handle: item.handle,
        path: path
      });
    }

    // Filtering the actions coming from dolman
    app._router.stack
      .reduce(reduceStack.bind(null, ''), [])
      .map(function(item) {
        return {
          route: routesMap.get(item.handle),
          path: item.path
        };
      })
      .filter(function(item) {
        return item.route && item.route.name;
      })
      .forEach(function(item) {
        var route = item.route,
            method = [].concat(route.method || route.methods)[0];

        var routeData = {
          path: item.path,
          name: route.name,
          method: !method || method === 'ALL' ? 'GET' : method
        };

        ['description'].forEach(function(k) {
          if (route[k])
            routeData[k] = route[k];
        });

        routes[route.name] = routeData;
      });

    return {
      formats: ['json'],
      methods: routes
    };
  }

  // Returning an object to handle
  return {
    router: makeRouter,
    specs: specs
  };
};


src/helpers.js
/**
 * Dolman Helpers
 * ===============
 *
 * Miscellaneous helper functions.
 */

/**
 * Is the target a plain object?
 */
function isPlainObject(target) {
  return target &&
         typeof target === 'object' &&
         !Array.isArray(target) &&
         !(target instanceof Date) &&
         !(target instanceof RegExp) &&
         !(target instanceof Map) &&
         !(target instanceof Set);
}

/**
 * Keep only some keys of an object.
 */
function keep(object, keys) {
  var output = {},
      k;

  keys = new Set(keys);

  for (k in object) {
    if (keys.has(k))
      output[k] = object[k];
  }

  return output;
}

/**
 * Unescaping a regex to retrieve the underlying path.
 */
function unescapeRegex(regex) {
  return regex.source
    .replace(/\?\(\?=\\\/\|\$\)/g, '')
    .replace(/^\^\\\//, '/')
    .replace(/\\\//g, '');
}

module.exports = {
  isPlainObject: isPlainObject,
  keep: keep,
  unescapeRegex: unescapeRegex
};


src/responses.js
/**
 * Dolman Responses
 * =================
 *
 * Overloading an express app's Response object to provide the user with handy
 * and semantic method to send back data to the client.
 */
var helpers = require('./helpers.js');

var isPlainObject = helpers.isPlainObject;

module.exports = function(app, logger, types) {
  var response = app.response;

  function applyMask(object, def) {
    var output = {},
        k,
        i,
        l;

    if (Array.isArray(def)) {
      if (!Array.isArray(object))
        return object;

      output = [];
      for (i = 0, l = object.length; i < l; i++) {
        output.push(applyMask(object[i], def[0]));
      }

      return output;
    }

    if (!isPlainObject(object) || !isPlainObject(def))
      return object;

    // TODO: optimize function call here?
    for (k in def) {
      if (k in object)
        output[k] = applyMask(object[k], def[k]);
    }

    return output;
  }

  function makeSuccess(code) {
    return function(result) {
      var maskedResult = result,
          def;

      // Do we need to apply a mask?
      if (this.__mask) {
        def = this.__mask;

        // We only keep relevant keys
        maskedResult = applyMask(result, def);

        // Then we check is the sent data does abide to the definition
        if (!types.check(def, maskedResult)) {
          logger.warn('Mask Warning', {
            mask: def,
            data: result
          });
        }
      }

      var data = {
        status: 'ok',
        code: code,
        result: maskedResult
      };

      if (this.__shouldBeCached)
        this.__sentData = maskedResult;

      return this.status(code).json(data);
    };
  }

  /**
   * Ok.
   */
  response.ok = makeSuccess(200);

  /**
   * Created.
   */
  response.created = makeSuccess(201);

  /**
   * Bad request.
   */
  response.badRequest = function(reason) {
    var data = {
      status: 'error',
      code: 400,
      error: {
        message: 'Bad Request'
      }
    };

    if (reason)
      data.error.reason = reason;

    return this.status(400).json(data);
  };

  /**
   * Unauthorized.
   */
  response.unauthorized = function() {
    this.status(401).json({
      status: 'error',
      code: 401,
      error: {
        message: 'Unauthorized'
      }
    });
  };

  /**
   * Forbidden.
   */
  response.forbidden = function() {
    this.status(403).json({
      status: 'error',
      code: 403,
      error: {
        message: 'Forbidden'
      }
    });
  };

  /**
   * Not Found.
   */
  response.notFound = function(reason) {
    var data = {
      status: 'error',
      code: 404,
      error: {
        message: 'Not Found'
      }
    };

    if (reason)
      data.error.reason = reason;

    this.status(404).json(data);
  };

  /**
   * Server Error.
   */
  response.serverError = function(err) {
    if (err)
      logger.error('Server Error', err);

    this.status(500).json({
      status: 'error',
      code: 500,
      error: {
        message: 'Internal Server Error'
      }
    });
  };
};


src/middlewares.js
/**
 * Dolman Middlewares
 * ===================
 *
 * Collection of handy middlewares.
 */

/**
 * Creating a RAM caching middleware.
 */
function cache(store, params) {
  if (typeof params === 'string')
    params = {key: params};

  var key = params.key,
      hasher = typeof params.hasher === 'function' ?
        params.hasher :
        function() {
          return '$nohash$';
        };

  // Initializing the store's key
  store[key] = {};

  return function(req, res, next) {
    var hash = hasher(req);

    // Do we have the data already?
    if (hash in store[key])
      return res.ok(store[key][hash]);

    // Flagging
    res.__shouldBeCached = true;

    // Catching response
    res.on('finish', function() {

      store[key][hash] = res.__sentData;
      delete res.__sentData;
    });

    return next();
  };
}

/**
 * Leveraging HTTP cache control.
 */
function httpCache(params) {
  var header = null;

  if (typeof params === 'string')
    header = params;

  // If params is an object, use special properties
  // defined by API to set `max-age`.
  if (params !== null && typeof params === 'function' || typeof params === 'object') {
    var durations = {
      seconds: 1, minutes: 60, hours: 3600, days: 86400, weeks: 604800
    };

    Object.keys(durations).some(function(key) {
      if (params.hasOwnProperty(key)) {
        header = 'private, max-age=' + durations[key] * params[key];
        return header;
      }
    });

    if (!header)
      throw new Error('Wrong parameter given for HTTP cache control');
  }

  return function(req, res, next) {
    res.set('Cache-Control', header);
    return next();
  };
}

/**
 * Applying a mask on output data.
 */
function mask(def) {
  return function(req, res, next) {
    res.__mask = def;
    return next();
  };
}

/**
 * Factory building a validation middleware working for the given definition.
 */
function validate(types, def) {
  return function(req, res, next) {

    var sources = ['params', 'query', 'body'],
        source,
        i,
        l;

    for (i = 0, l = sources.length; i < l; i++) {
      source = sources[i];

      if (def[source] && !types.check(def[source], req[source])) {
        var reason = {
          source: source,
          expecting: def[source],
          sent: req[source] || {}
        };

        if (source === 'params')
          reason.path = req.route.path;

        return res.badRequest(reason);
      }
    }

    return next();
  };
}

module.exports = {
  cache: cache,
  httpCache: httpCache,
  mask: mask,
  validate: validate
};
